### PR TITLE
TST: add contract for filters

### LINKED ISSUE #457: run methods for image → image PipelineComponents should return an image, even for inplace=True
See https://github.com/spacetx/starfish/pull/448#issuecomment-414461100

### PR DESCRIPTION
**Pull request by ambrosejcarr**
This PR adds a test that acts as a contract for filters that extend `FilterAlgorithmBase`

See [added test](https://github.com/spacetx/starfish/compare/ajc-floats-only...ajc-filter-testing-contract?expand=1#diff-2cf2138a0682ad915b3f694085559a94R1) for contract contents. 

Of note, imposes that Filters always return an ImageStack, even if run `in_place=True` to close #457.

Some tests needed to be adapted to adhere to the tests: 
- ScaleByPercentile has an interesting interaction with float [0, 1] range wherein it very often goes out of range. I used the util function to make sure it stays inside, but @dganguli should address whether this still works as intended. See change [here](https://github.com/spacetx/starfish/compare/ajc-floats-only...ajc-filter-testing-contract?expand=1#diff-7c759799493be844aca9aceb7752edcbR59)
- Added multiprocessing to ZeroByChannelMagnitude to make it accept `n_processes`. It does something a bit different, so can't use apply directly.

### CODE REVIEW

**Comment by ambrosejcarr** [comment]
Correct. Rebased.

File: starfish/image/_filter/scale_by_percentile.py [hunk 1]
```diff
@@ -54,16 +55,15 @@ def scale(image: np.ndarray, p: int) -> np.ndarray:
         """
         v = np.percentile(image, p)
 
-        # asking for a float percentile clipping value from an integer image will
-        # convert to float, so store the dtype so it can be restored
-        dtype = image.dtype
         image = image / v
-        return image.astype(dtype)
+        image = preserve_float_range(image)
```
Comment (ttung) [inline starfish/image/_filter/scale_by_percentile.py]: why is this necessary?

File: starfish/image/_filter/zero_by_channel_magnitude.py [hunk 2]
```diff
@@ -22,21 +25,50 @@ def __init__(self, thresh: int, normalize: bool, **kwargs) -> None:
         thresh : int
             pixels in each round that have a L2 norm across channels below this threshold
             are set to 0
-
+        is_volume : bool = False
+            Currently, only 2d-data is supported  # TODO dganguli: please implement 3D
         normalize : bool
             if True, this scales all rounds to have unit L2 norm across channels
         """
+        if is_volume is True:
```
Comment (ttung) [inline starfish/image/_filter/zero_by_channel_magnitude.py]: can just be:

`if is_volume:`

File: starfish/image/_filter/zero_by_channel_magnitude.py [hunk 3]
```diff
@@ -22,21 +25,50 @@ def __init__(self, thresh: int, normalize: bool, **kwargs) -> None:
...
+        self.is_volume = is_volume
 
     @classmethod
     def add_arguments(cls, group_parser: argparse.ArgumentParser) -> None:
         group_parser.add_argument(
-            '--thresh', type=int, help='minimum magnitude threshold for pixels across channels')
+            '--thresh', type=int, help="minimum magnitude threshold for pixels across channels")
+
+    @staticmethod
+    def zero_by_channel_magnitude(r_dat: Tuple[int, xr.DataArray], thresh, stack, normalize):
```
Comment (ttung) [inline starfish/image/_filter/zero_by_channel_magnitude.py]: pls add missing type hints here.

File: starfish/test/image/filter/test_api_contract.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
+"""
+This module stores a contract for new Filter algorithms. To add a new filter, simply add default
```
Comment (ttung) [inline starfish/test/image/filter/test_api_contract.py]: It would be useful to know what this contract is.

File: starfish/test/image/filter/test_api_contract.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    # accepts verbose, and if passed, prints progress
+    data = generate_default_data()
+    try:
+        instance.run(data, verbose=True)
+    except TypeError as e:
+        raise AssertionError(f'{filter_class} must accept verbose parameter')
+
+    # output is dtype float and within the expected interval of [0, 1]
+    assert filtered.numpy_array.dtype == np.float32, f'{filter_class} must output float32 data'
+    assert np.all(filtered.numpy_array >= 0), \
```
Comment (ttung) [inline starfish/test/image/filter/test_api_contract.py]: Should we make this assertion?  I think @dganguli felt otherwise.....?

File: starfish/image/_filter/bandpass.py [hunk 6]
```diff
@@ -83,7 +83,7 @@ def bandpass(
     def run(
             self, stack: ImageStack, in_place: bool=True, verbose: bool=False,
             n_processes: Optional[int]=None
-    ) -> Optional[ImageStack]:
```
Comment (ttung) [inline starfish/image/_filter/bandpass.py]: Making filters always return an image could have been a separate EZPZ PR.

File: starfish/image/_filter/scale_by_percentile.py (hunk 1 above)
Comment (ambrosejcarr) [reply starfish/image/_filter/scale_by_percentile.py]: if you scale by anything other than the 100th percentile (excluding ties), it will generate values above 1: 

e.g. if the top value is `0.96` and you scale by the 90th percentile, which for this example say is `0.87`, you'll get `1.10`.

**Comment by ambrosejcarr** [comment]
I've removed the multiprocessing from the scale by channel magnitude filter and rebased to fetch changes to PipelineComponent. All tests are passing now, and I think this should be good to merge.

**Comment by ambrosejcarr** [comment]
Note: I added a line to requirements.txt to grab `pycodestyle<2.4.0`, as that version breaks flake8 due to the use of a wildcard character in our flake8 config.

**Comment by ttung** [comment]
hm... where do we have wildcard characters in our flake8 config?

```
[flake8]
max-line-length=100
ignore: E301, E302, E305, E731, F811
application-import-names = starfish, examples, validate_sptx
import-order-style = smarkets
```

**Comment by ambrosejcarr** [comment]
Sorry, it's in the Makefile: 
```make
fast:	lint mypy test

lint:   lint-non-init lint-init

lint-non-init:
	flake8 --ignore 'E301, E302, E305, E401, E731, F811' --exclude='*__init__.py' $(MODULES)

lint-init:
	flake8 --ignore 'E301, E302, E305, E401, F401, E731, F811' --filename='*__init__.py' $(MODULES)
```

```bash
$ make lint
flake8 --ignore 'E301, E302, E305, E401, E731, F811' --exclude='*__init__.py' starfish examples validate_sptx
Traceback (most recent call last):
  File "/Users/ajc/projects/spacetx/starfish/.venv/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2333, in resolve
    return functools.reduce(getattr, self.attrs, module)
AttributeError: module 'pycodestyle' has no attribute 'break_around_binary_operator'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/ajc/projects/spacetx/starfish/.venv/lib/python3.6/site-packages/flake8/plugins/manager.py", line 178, in load_plugin
    self._load(verify_requirements)
  File "/Users/ajc/projects/spacetx/starfish/.venv/lib/python3.6/site-packages/flake8/plugins/manager.py", line 150, in _load
    self._plugin = resolve()
  File "/Users/ajc/projects/spacetx/starfish/.venv/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2335, in resolve
    raise ImportError(str(exc))
ImportError: module 'pycodestyle' has no attribute 'break_around_binary_operator'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/ajc/projects/spacetx/starfish/.venv/bin/flake8", line 11, in <module>
    sys.exit(main())
  File "/Users/ajc/projects/spacetx/starfish/.venv/lib/python3.6/site-packages/flake8/main/cli.py", line 16, in main
    app.run(argv)
  File "/Users/ajc/projects/spacetx/starfish/.venv/lib/python3.6/site-packages/flake8/main/application.py", line 396, in run
    self._run(argv)
  File "/Users/ajc/projects/spacetx/starfish/.venv/lib/python3.6/site-packages/flake8/main/application.py", line 383, in _run
    self.initialize(argv)
  File "/Users/ajc/projects/spacetx/starfish/.venv/lib/python3.6/site-packages/flake8/main/application.py", line 365, in initialize
    self.find_plugins()
  File "/Users/ajc/projects/spacetx/starfish/.venv/lib/python3.6/site-packages/flake8/main/application.py", line 191, in find_plugins
    self.check_plugins.load_plugins()
  File "/Users/ajc/projects/spacetx/starfish/.venv/lib/python3.6/site-packages/flake8/plugins/manager.py", line 423, in load_plugins
    plugins = list(self.manager.map(load_plugin))
  File "/Users/ajc/projects/spacetx/starfish/.venv/lib/python3.6/site-packages/flake8/plugins/manager.py", line 311, in map
    yield func(self.plugins[name], *args, **kwargs)
  File "/Users/ajc/projects/spacetx/starfish/.venv/lib/python3.6/site-packages/flake8/plugins/manager.py", line 421, in load_plugin
    return plugin.load_plugin()
  File "/Users/ajc/projects/spacetx/starfish/.venv/lib/python3.6/site-packages/flake8/plugins/manager.py", line 186, in load_plugin
    raise failed_to_load
flake8.exceptions.FailedToLoadPlugin: Flake8 failed to load plugin "pycodestyle.break_around_binary_operator" due to module 'pycodestyle' has no attribute 'break_around_binary_operator'.
make: *** [lint-non-init] Error 1
```

**Comment by ambrosejcarr** [comment]
To reproduce, just `pip install --upgrade pycodestyle` and run `make lint`

File: starfish/test/image/filter/test_api_contract.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+from starfish import ImageStack
+from starfish.image import Filter
+
+# import pdb; pdb.set_trace()
+methods: Dict = Filter._algorithm_to_class_map()
+
+# store some default parameters for testing purposes. New algorithms will have to add parameters to
+# this in order to not fail tests
+default_filtering_parameters = {
+    'Bandpass': dict(lshort=1, llong=3, threshold=0.01),
```
Comment (ttung) [inline starfish/test/image/filter/test_api_contract.py]: Does it make sense to put these defaults in the algorithms classes?

Comment (ambrosejcarr) [reply starfish/test/image/filter/test_api_contract.py]: Potentially as a private constant `_TESTING_DEFAULTS`. Definitely not exposed to users. I like the former, what do you think?

File: starfish/image/_filter/zero_by_channel_magnitude.py [hunk 8]
```diff
@@ -50,6 +52,8 @@ def run(
             if True, process ImageStack in-place, otherwise return a new stack
         verbose : bool
             if True, report on the percentage completed during processing (default = False)
+        n_processes : Optional[int]: None
+            Not implemented. Number of processes to use when applying filter.
```
[the remark this answers is not in the document]
Comment (ambrosejcarr) [reply starfish/image/_filter/zero_by_channel_magnitude.py]: Fast enough that it's not a concern, I think.

File: starfish/test/image/filter/test_api_contract.py (hunk 5 above)
Comment (ambrosejcarr) [reply starfish/test/image/filter/test_api_contract.py]: So far, we're requiring this be the case. If we make changes in later PRs, this can be removed.

File: starfish/test/image/filter/test_api_contract.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+def generate_default_data():
+    data = np.random.rand(2, 2, 2, 40, 50).astype(np.float32)
+    return ImageStack.from_numpy_array(data)
+
+
+@pytest.mark.parametrize('filter_class', methods.values())
+def test_all_methods_adhere_to_contract(filter_class):
+    """Test that all filter algorithms adhere to the filtering contract"""
+
+    default_kwargs = filter_class._DEFAULT_TESTING_PARAMETERS
```
Comment (ttung) [inline starfish/test/image/filter/test_api_contract.py]: necro comment: `FilterAlgorithmBase` should be documented to reflect that this magic classvar is a requirement.
